File: test/e2e/Apps/BasicDotNetIsolated/LargeOutputOrchestrator.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask;
using Microsoft.DurableTask.Client;
using Microsoft.Extensions.Logging;
using System.Net;
using System.Text.Json.Nodes;

namespace Microsoft.Azure.Durable.Tests.E2E;

public static class LargeOutputOrchestrator
{
    [Function(nameof(LargeOutputOrchestrator))]
    public static async Task<List<string>> RunOrchestrator(
        [OrchestrationTrigger] TaskOrchestrationContext context)
    {
        ILogger logger = context.CreateReplaySafeLogger(nameof(LargeOutputOrchestrator));
        int sizeInKB = context.GetInput<int>();

        logger.LogInformation("Saying hello.");
        var outputs = new List<string>();

        outputs.Add(await context.CallActivityAsync<string>(nameof(LargeOutputSayHello), "Tokyo"));

        // Add a large message to the outputs that exceeds the Azure Storage Queue message size limit (64 KB),
        // so that blobs will be used instead. 
        outputs.Add(GenerateLargeString(sizeInKB));

        return outputs;
    }

    [Function(nameof(LargeOutputSayHello))]
    public static string LargeOutputSayHello([ActivityTrigger] string name, FunctionContext executionContext)
    {
        return $"Hello {name}!";
    }

    [Function("LargeOutputOrchestrator_HttpStart")]
    public static async Task<HttpResponseData> HttpStart(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client,
        FunctionContext executionContext)
    {
        ILogger logger = executionContext.GetLogger("LargeOutputOrchestrator_HttpStart");
        int sizeInKB = await req.ReadFromJsonAsync<int>();
        
        // Function input comes from the request content.
        string instanceId = await client.ScheduleNewOrchestrationInstanceAsync(
            nameof(LargeOutputOrchestrator), input: sizeInKB);

        logger.LogInformation("Started orchestration with ID = '{instanceId}'.", instanceId);

        // Returns an HTTP 202 response with an instance management payload.
        // See https://learn.microsoft.com/azure/azure-functions/durable/durable-functions-http-api#start-orchestration
        return await client.CreateCheckStatusResponseAsync(req, instanceId);
    }

    [Function("LargeOutputOrchestrator_Query_Output")]
    public static async Task<HttpResponseData> QueryOutput(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
        [DurableClient] DurableTaskClient client,
        string id)
    {
        OrchestrationMetadata? metadata = await client.GetInstancesAsync(instanceId: id, getInputsAndOutputs:true);
   
        HttpResponseData response;
        if (metadata == null)
        {
            response = req.CreateResponse(HttpStatusCode.NotFound);
            await response.WriteStringAsync("Orchestration metadata not found.");
            return response; // Return a 404 response if metadata is null
        }

        var output = metadata.ReadOutputAs<JsonArray>();

        response = req.CreateResponse(HttpStatusCode.OK);
        response.Headers.Add("Content-Type", "application/json");
        await response.WriteStringAsync(output!.ToString());

        return response;
    }

    static string GenerateLargeString(int sizeInKB)
    {
        return new string('A', sizeInKB * 1024);
    }
}

File: test/e2e/Apps/BasicJava/src/main/java/com/function/LargeOutputOrchestrator.java
package com.function;

import com.microsoft.azure.functions.annotation.*;
import com.function.JsonHelpers.DurableMetadataGsonProvider;
import com.google.gson.Gson;
import com.microsoft.azure.functions.*;

import java.util.*;

import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableActivityTrigger;
import com.microsoft.durabletask.azurefunctions.DurableClientContext;
import com.microsoft.durabletask.azurefunctions.DurableClientInput;
import com.microsoft.durabletask.azurefunctions.DurableOrchestrationTrigger;

/**
 * LargeOutputOrchestrator Java implementation.
 */
public class LargeOutputOrchestrator {

    @FunctionName("LargeOutputOrchestrator")
    public List<String> runOrchestrator(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx, 
            final ExecutionContext context) {
        int sizeInKB = ctx.getInput(Integer.class);

        context.getLogger().info("Saying hello.");
        List<String> outputs = new ArrayList<>();

        outputs.add(ctx.callActivity("LargeOutputSayHello", "Tokyo", String.class).await());

        // Add a large message to the outputs that exceeds the Azure Storage Queue message size limit (64 KB),
        // so that blobs will be used instead.
        outputs.add(generateLargeString(sizeInKB));

        return outputs;
    }

    @FunctionName("LargeOutputSayHello")
    public String largeOutputSayHello(
            @DurableActivityTrigger(name = "name") String name,
            final ExecutionContext context) {
        return "Hello " + name + "!";
    }

    @FunctionName("LargeOutputOrchestrator_HttpStart")
    public HttpResponseMessage httpStart(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            final ExecutionContext context) {
        int sizeInKB = 0;
        Gson gson = DurableMetadataGsonProvider.createGson();
        try {
            String body = request.getBody().orElse("0");
            sizeInKB = gson.fromJson(body, int.class);
        } catch (Exception e) {
            sizeInKB = 0;
        }

        DurableTaskClient client = durableContext.getClient();
        String instanceId = client.scheduleNewOrchestrationInstance("LargeOutputOrchestrator", sizeInKB);
        context.getLogger().info("Started orchestration with ID = '" + instanceId + "'.");

        return durableContext.createCheckStatusResponse(request, instanceId);
    }

    @FunctionName("LargeOutputOrchestrator_Query_Output")
    public HttpResponseMessage queryOutput(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS, route = "LargeOutputOrchestrator_Query_Output") 
            HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") 
            DurableClientContext durableContext,
            final ExecutionContext context) {
        DurableTaskClient client = durableContext.getClient();
        String instanceId = request.getQueryParameters().get("id");
        OrchestrationMetadata metadata = client.getInstanceMetadata(instanceId, true);

        if (metadata == null) {
            HttpResponseMessage response = request.createResponseBuilder(HttpStatus.NOT_FOUND)
                .body("Orchestration metadata not found.").build();
            return response;
        }

        Gson gson = DurableMetadataGsonProvider.createGson();
        String outputJson = gson.toJson(metadata);

        HttpResponseMessage response = request.createResponseBuilder(HttpStatus.OK)
            .header("Content-Type", "application/json")
            .body(outputJson)
            .build();
        return response;
    }

    private static String generateLargeString(int sizeInKB) {
        char[] chars = new char[sizeInKB * 1024];
        Arrays.fill(chars, 'A');
        return new String(chars);
    }
}